Working expression optimization, and some improvements to branch-level source coverage

richkadel opened the pull request:
This replaces PR #78040 after reorganizing the original commits (by request) into a more logical sequence of major changes.

Most of the work is in the MIR `transform/coverage/` directory (originally, `transform/instrument_coverage.rs`). 

Note this PR includes some significant additional debugging capabilities, to help myself and any future developer working on coverage improvements or issues.

In particular, there's a new Graphviz (.dot file) output for the coverage graph (the `BasicCoverageBlock` control flow graph) that provides ways to get some very good insight into the relationships between the MIR, the coverage graph BCBs, coverage spans, and counters. (There are also some cool debugging options, available via environment variable, to alter how some data in the graph appears.)

And the code for this Graphviz view is actually generic... it can be used by any implementation of the Rust `Graph` traits.

Finally (for now), I also now output information from `llvm-cov` that shows the actual counters and spans it found in the coverage map, and their counts (from the `--debug` flag). I found this to be enormously helpful in debugging some coverage issues, so I kept it in the test results as well for additional context.

@tmandry @wesleywiser 

r? @tmandry 

Here's an example of the new coverage graph:

* Within each `BasicCoverageBlock` (BCB), you can see each `CoverageSpan` and its contributing statements (MIR `Statement`s and/or `Terminator`s)
* Each `CoverageSpan` has a `Counter` or and `Expression`, and `Expression`s show their Add/Subtract operation with nested operations. (This can be changed to show the Counter and Expression IDs instead, or in addition to, the BCB.)
* The terminators of all MIR `BasicBlock`s in the BCB, including one final `Terminator`
* If an "edge counter" is required (because we need to count an edge between blocks, in some cases) the edge's Counter or Expression is shown next to its label. (Not shown in the example below.) (FYI, Edge Counters are converted into a new MIR `BasicBlock` with `Goto`)

<img width="1116" alt="Screen Shot 2020-10-17 at 12 23 29 AM" src="https://user-images.githubusercontent.com/3827298/96331095-616cb480-100f-11eb-8212-60f2d433e2d8.png">

r? @tmandry 
FYI: @wesleywiser

richkadel commented:
@tmandry - I reorganized the commits in the most logical way I could come up with. Does this work for you?

If so, this is ready for your review, and other than feedback changes, it's ready to merge, as far as I'm concerned. (No more changes planned, to this PR.)

@wesleywiser - Just FYI, I'm taking Mon-Wed off next week so I may not respond right away, after Saturday, if you are sending feedback as well.

Thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #78319) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

richkadel commented:
I resolved the conflicts introduced by my bugfix PR #78300

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -54,69 +50,131 @@ pub(crate) fn provide(providers: &mut Providers) {
+/// safeguard, with `add_missing_operands` set to `true`, to find any other counter or expression
+/// IDs referenced by expression operands, if not already seen.
+///
+/// Ideally, every expression operand in the MIR will have a corresponding Counter or Expression,
```
```suggestion
/// Ideally, every counter expression id operand in the MIR will have a corresponding Counter or Expression,
```

richkadel replied:
Point taken, but I reworded it a little differently. Let me know if it's better or not.

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -54,69 +50,131 @@ pub(crate) fn provide(providers: &mut Providers) {
+    // If an expression operand is encountered with an ID outside the range of known counters and
+    // expressions, the only way to determine if the ID is a counter ID or an expression ID is to
+    // assume a maximum possible counter ID value.
+    const MAX_COUNTER_GUARD: u32 = (u32::MAX / 2) + 1;
```
Do we check that we don't go beyond this anywhere?

richkadel replied:
No, so I added this check (in `counters.rs`)

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -54,69 +50,131 @@ pub(crate) fn provide(providers: &mut Providers) {
+    coverage_visitor.visit_body(mir_body);
+
     coverage_visitor.info
 }
```
```suggestion
}

```

richkadel replied:
I think `x.py fmt` did this, but no matter. The line changed in a subsequent commit, and as a result, it has this multiline formatting now.

richkadel replied:
This is split into different files at this line now. (I probably joined them when reconstructing the commits in reverse, sorry.)

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -54,69 +50,131 @@ pub(crate) fn provide(providers: &mut Providers) {
+/// This visitor runs twice, first with `add_missing_operands` set to `false`, to find the maximum
+/// counter ID and maximum expression ID based on their enum variant `id` fields; then, as a
+/// safeguard, with `add_missing_operands` set to `true`, to find any other counter or expression
+/// IDs referenced by expression operands, if not already seen.
```
What keeps us from doing this in one pass? If we're falling back to the MAX_COUNTER_GUARD anyway it seems like we could always use that?

richkadel replied:
I need to revisit this when I get back. My first thought was, no because I'm only forcing the counters to be less than the max, but expressions could be larger than u32:MAX minus the max counters. But if that happens then counters really can't reach the max counters constant so using that probably won't work anyway. That means I would have to actually pick a max value and stick to it for both counters and expressions, which I don't really like because I don't know what the right split should be, and it could change over time. If I can't come up with a way to dynamically figure this out like I used to, then I think I may just have to do that, in which case yes maybe I can make it just one pass here.

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -589,126 +640,212 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
+        self.num_expressions += 1;
+        InjectedExpressionId::from(next)
     }
+}
```
```suggestion
}

```

richkadel replied:
This is no longer structured this way. Probably accidentally joined when restructuring the commits in reverse.

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -839,38 +941,72 @@ impl<'a> CoverageSpanRefinery<'a> {
+        bcb.blocks
+            .iter()
+            .map(|bbref| {
+                let bb = *bbref;
```
nit: I think this would work
```suggestion
            .map(|ref bb| {
```

richkadel replied:
That didn't work, but `|&bb|` does.

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -839,38 +941,72 @@ impl<'a> CoverageSpanRefinery<'a> {
+                            .map(|span| CoverageSpan::for_terminator(span, bcb, bb)),
+                    )
+            })
+            .flatten()
```
nit: I think you can use `flat_map` instead of `map` then `flatten`?

richkadel replied:
👍 Done

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -1138,15 +1302,43 @@ fn filtered_terminator_span(terminator: &'a Terminator<'tcx>, body_span: Span) -
+        // `FalseEdge`.
         | TerminatorKind::FalseEdge { .. } => None,
 
+        // FIXME(richkadel): Note that `Goto` was initially filtered out (by returning `None`, as
```
Might be worth opening an issue to track all the kinds of "span weirdness" encountered and collect FIXMEs like this. I'd like to see some discussion about what we could change in MIR building / other upstream code.

richkadel replied:
Got it. I'll create an issue for this (and look for others to add) and add Issue references in the code once they are created.

richkadel replied:
Issue #78542 (I'm adding this to the comment)

tmandry commented on compiler/rustc_mir/src/transform/instrument_coverage.rs:
```diff
@@ -1138,15 +1302,43 @@ fn filtered_terminator_span(terminator: &'a Terminator<'tcx>, body_span: Span) -
+        // However, in other cases, a visible `CoverageSpan` is not wanted, but the `Goto`
+        // block must still be counted (for example, to contribute its count to an `Expression`
+        // that reports the execution count for some other block). In these cases, the code region
+        // is set to `None`.
```
where?

richkadel replied:
Reference added.

tmandry reviewed (commented):
Just got through the first commit today.

I don't remember if I've asked this before, but wanted to check that unwind edges still split BCBs. I know we don't add counters for them, but we still need to make sure that code after a call isn't counted if the call panics.

Is there a way we can only add the html test output if it's really adding something to the test? I suspect it's redundant in many cases. I'm certainly not reading through it in each diff :)

Finally, food for thought: Could some of the things checked by these end-to-end tests actually be checked by unit tests in the code? Unit tests aren't too common in the compiler today, but with as much intricate logic as there is in this code, I think they're worth considering.

richkadel commented:
> I don't remember if I've asked this before, but wanted to check that unwind edges still split BCBs. I know we don't add counters for them, but we still need to make sure that code after a call isn't counted if the call panics.

I'm adding a second `call()` right after the first `call(true)` in the test `try_error_result.rs`, to show this works for managed errors:

https://github.com/rust-lang/rust/blob/a1dfd2490a6cb456b92e469fa550dc217e20ad6d/src/test/run-make-fulldeps/coverage/try_error_result.rs#L24-L32

Here's the result:

<img width="513" alt="Screen Shot 2020-10-24 at 10 03 54 PM" src="https://user-images.githubusercontent.com/3827298/97099311-e45fc180-1644-11eb-83fb-7e37b5205c19.png">

The coverage results are exactly what I would expect: The `?` after `call(true)` (which returns `Err`) is not highlighted in red here, which shows the error handler was executed (1 time; the same number of times as the call itself). The `call(false)` immediately after that (before `else`) is not covered (the entire line is red, and the count is `0`), which is correct.

I remember discussing unmanaged errors a while back, and here's my position again (which I still think is right): For the initial release, if a program panics, I believe the results will probably be the same results one might expect for a clang program that crashes: that is, they are not guaranteed to be accurate, and I'm not making any guarantees.

If someone thinks we should support accurate coverage results for programs that panic, we can make it a feature request. 

For now, just trying this in the `try_error_result.rs` test, replacing the `Err()` return status with `panic!()`, the result they get is (per my current approach):

```bash
$ LLVM_PROFILE_FILE=try_error_result.profraw ./try_error_result
thread 'main' panicked at 'explicit panic', src/test/run-make-fulldeps/coverage/try_error_result.rs:6:9
stack backtrace:
   0:     0x5608e9f59e31 - std::backtrace_rs::backtrace::libunwind::trace::he2381431404d861d
                               at /usr/local/google/home/richkadel/rust/library/std/src/../../backtrace/src/backtrace/libunwind.rs:100:5
```

To your question, I don't inject separate counters for each function call in case of unwind, so depending on where the counter is incremented, it may show all non-branching statements (including function calls) as having been executed, or non of them, even thought a panic!() may interrupt the sequence somewhere in the middle.

It would be inefficient to inject separate counters and coverage regions for each call, just in case one of them panics. It could be done, but I don't think it's worth doing in this first version.

That's still my position.

Are you OK with that for this release?

richkadel commented:
> Is there a way we can only add the html test output if it's really adding something to the test? I suspect it's redundant in many cases. I'm certainly not reading through it in each diff :)

No I hope not! 😄  I wouldn't expect anyone to do that. I really intended for these to be used in two ways:

1. If someone changes the MIR in a way that breaks how the CoverageSpans are computed (which could be a big problem), they can compare the differences and resolve the problem.
2. One can review the CoverageSpans in the spanview rendered format. Reviewing the actual source is way too painful. I hardly ever do that myself, and it's not nearly as helpful as reviewing the rendered form.

I was thinking about inserting a comment at the top of the file with a link to view the file in the browser, using `htmlpreview.github.io`. I think I'll do that in the next round of fixes based on your comments.

BTW, I was actually also thinking it would be nice to review the new CoverageGraph too. (It would have answered your question about unwind, in a way, for example.)

I could still do that. It would mean adding even more files not intended for review "as source", but the only wan to review the rendered version is (AFAICT) copy it into a local file, or download the patch, and then open it in a graphviz viewer, such as the VSCode plugin I use.

Not sure it's worth it, but it wouldn't be hard to add.

WDYT?

richkadel commented:
> Finally, food for thought: Could some of the things checked by these end-to-end tests actually be checked by unit tests in the code? Unit tests aren't too common in the compiler today, but with as much intricate logic as there is in this code, I think they're worth considering.

Yes, it's a good idea. I've thought about it too. I'm not against it, but the existing tests have worked pretty well so far, so I haven't felt that motivated. Maybe something for the future?

richkadel commented on compiler/rustc_codegen_ssa/src/coverageinfo/map.rs:
```diff
@@ -129,54 +156,60 @@ impl FunctionCoverage {
+        // `expression_index`s lower than the referencing `Expression`. Therefore, it is
         // reasonable to look up the new index of an expression operand while the `new_indexes`
         // vector is only complete up to the current `ExpressionIndex`.
         let id_to_counter =
```
Added more comments. Hopefully this helps?

richkadel commented:
> I was thinking about inserting a comment at the top of the file with a link to view the file in the browser, using `htmlpreview.github.io`. I think I'll do that in the next round of fixes based on your comments.

@tmandry - I'm still working through your feedback, but I wanted to check in a partial response to the first couple of comments; particularly the addition of a link (URL) to a rendered version of the Spanview. (See my comments above.)

See for example:

https://github.com/rust-lang/rust/blob/c3cb945ff97134bf5abac4bedf3a97ffa91a815f/src/test/run-make-fulldeps/coverage-spanview-base/expected_mir_dump.closure/closure.main.-------.InstrumentCoverage.0.html

Select the URL and open it in your browser. It should work for simple function names, but I do have to fix the URL for special characters, and escape things like curly braces. (I just realized that.)

richkadel commented:
URLs are now escaped:

See the header comment in the following, for example:

https://github.com/rust-lang/rust/blob/64aaaedc56b5d379443c1716224ec20ee95993b3/src/test/run-make-fulldeps/coverage-spanview-base/expected_mir_dump.closure/closure.main-%7Bclosure%230%7D.-------.InstrumentCoverage.0.html

richkadel commented on compiler/rustc_middle/src/mir/coverage.rs:
```diff
@@ -64,21 +85,35 @@ impl From<CounterValueReference> for ExpressionOperandId {
 }
 
-impl From<InjectedExpressionIndex> for ExpressionOperandId {
+impl From<&mut CounterValueReference> for ExpressionOperandId {
```
I swear I needed it at one time, but maybe some other change in my code elsewhere removed that requirement. Seems to work without it again.

richkadel replied:
Removed. Done.

richkadel commented on compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs:
```diff
@@ -10,7 +10,15 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+                    // expressions.
+                    bx.add_coverage_counter(self.instance, function_source_hash, id, code_region)
+                } else {
+                    bx.set_function_source_hash(self.instance, function_source_hash)
```
Done

richkadel commented:
@tmandry Thanks for this first review! I think I've addressed all of the actionable code comments. PTAL.

tmandry commented on compiler/rustc_mir/src/transform/coverage/counters.rs:
```diff
@@ -95,6 +105,7 @@ impl CoverageCounters {
     /// Counter IDs start from one and go up.
     fn next_counter(&mut self) -> CounterValueReference {
         assert!(self.next_counter_id < u32::MAX - self.num_expressions);
+        assert!(self.next_counter_id <= MAX_COUNTER_GUARD);
```
You have `+ 1` above and `<=` here. I guess it doesn't really matter but I'd expect one or the other, not both.

richkadel replied:
Agreed, and at first I did remove the `+ 1` from the constant def, but (as I will explain below) I believe I'm able to remove `MAX_COUNTER_GUARD` completely, which is even better. So that second `assert!` is gone as well.

richkadel commented:
Making it optional is probably the way to go. I will try to file an issue
for this, along with the other issue or issues discussed in this set of
review comments.

On Tue, Oct 27, 2020, 3:58 PM Tyler Mandry <notifications@github.com> wrote:

> Thanks for your replies. Regarding panics, I'm not sure what the right
> thing to do is. But I'm pretty sure there are some cases where we'll want
> to see correct results in the face of panics, namely #[should_panic]
> tests, and maybe in code that uses catch_unwind.
>
> Since these are a small percentage of use cases I'm fine with not
> supporting it just yet, but I think we'll at least want an option to turn
> on support at some point.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/78267#issuecomment-717589286>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5GMYR7LGDST5YUEIE5X7LSM5GAFANCNFSM4S4I3RRA>
> .
>

tmandry commented on compiler/rustc_mir/src/transform/coverage/debug.rs:
```diff
@@ -1,12 +1,294 @@
+}
+
+impl DebugOptions {
+    fn new() -> Self {
```
`from_env` or similar might be a clearer name than `new`

richkadel replied:
Done

tmandry commented on compiler/rustc_mir/src/transform/coverage/debug.rs:
```diff
@@ -1,12 +1,294 @@
+        self.some_counters.replace(FxHashMap::default());
+    }
+
+    pub fn is_enabled(&mut self) -> bool {
```
Doesn't need mut

richkadel replied:
Agree. I found a few other functions in `debug.rs` with the same issue and remove `mut` in all of them that didn't need it. (Copy-paste issue I think.)

tmandry commented on compiler/rustc_mir/src/transform/coverage/debug.rs:
```diff
@@ -1,12 +1,294 @@
+        Self { some_counters: None }
+    }
+
+    pub fn enable(&mut self) {
```
Might want to assert that it's disabled so you aren't replacing existing data.

richkadel replied:
Great catch! 

I added the following to all `enable()` functions in `debug.rs`:

```rust
        debug_assert!(!self.is_enabled());
```

tmandry reviewed (commented):
Just one commit left to go..

richkadel commented:
> I'm pretty sure there are some cases where we'll want to see correct results in the face of panics, namely `#[should_panic]` tests, and maybe in code that uses `catch_unwind`.
> 
> Since these are a small percentage of use cases I'm fine with not supporting it just yet, but I think we'll at least want an option to turn on support at some point.

I created Issue #78544 for this.

wesleywiser commented on compiler/rustc_mir/src/transform/coverage/debug.rs:
```diff
@@ -1,12 +1,294 @@
-use super::graph::CoverageGraph;
+use super::graph::{BasicCoverageBlock, BasicCoverageBlockData, CoverageGraph};
```
It would be nice to have a module level doc comment that explains what the debugging options available are and what they do.

richkadel replied:
Done! Check out the new module comments in `debug.rs` when I upload the next set of changes, this afternoon, and let me know what you think.

richkadel commented:
> > Finally, food for thought: Could some of the things checked by these end-to-end tests actually be checked by unit tests in the code? Unit tests aren't too common in the compiler today, but with as much intricate logic as there is in this code, I think they're worth considering.
> 
> Yes, it's a good idea. I've thought about it too. I'm not against it, but the existing tests have worked pretty well so far, so I haven't felt that motivated. Maybe something for the future?

Just to capture this idea, I added a short FIXME statement in `src/test/run-make-fulldeps/coverage/coverage_tools.mk`

tmandry commented on compiler/rustc_mir/src/transform/coverage/graph.rs:
```diff
@@ -316,6 +329,91 @@ impl BasicCoverageBlockData {
+            }
+        }
+        let operand = counter_kind.as_operand_id();
+        let expect_none = self
```
nit: can you use if let here?

richkadel replied:
Yep. I thought I was making it more readable this way, but after restructuring it again, with the `if let` this time, it actually seems better. Thanks.

tmandry commented on compiler/rustc_mir/src/transform/coverage/graph.rs:
```diff
@@ -316,6 +329,91 @@ impl BasicCoverageBlockData {
+        self.counter_kind.take()
+    }
+
+    #[inline(always)]
```
should all of these really be `#[inline(always)]`? some methods seem pretty big for it.

richkadel replied:
No, my bad. I removed the directive from the larger functions, and removed `(always)` from a few of them as well.

tmandry commented on compiler/rustc_mir/src/transform/coverage/graph.rs:
```diff
@@ -344,6 +492,175 @@ fn bcb_filtered_successors<'a, 'tcx>(
+/// CoverageGraph outside all loops. This supports traversing the BCB CFG in a way that
+/// ensures a loop is completely traversed before processing Blocks after the end of the loop.
+#[derive(Debug)]
+pub(crate) struct TraversalContext {
```
Another piece of code I think would benefit from some unit tests

richkadel replied:
I added a FIXME comment to add unit tests here as well.

tmandry commented on compiler/rustc_mir/src/transform/coverage/graph.rs:
```diff
@@ -344,6 +492,175 @@ fn bcb_filtered_successors<'a, 'tcx>(
+    let mut backedges = IndexVec::from_elem_n(Vec::<BasicCoverageBlock>::new(), num_bcbs);
+
+    // Identify loops by their backedges
+    for (bcb, _) in basic_coverage_blocks.iter_enumerated() {
```
What's the complexity of this? Thinking about how it would handle very large functions

richkadel replied:
Since it evoked your question, I added a comment in the code for this. I took a bit of an educated guess at the big-O complexity equation (not something I feel super comfortable with), so please read my comment and let me know if you would state it differently.

Bottom line, I think the complexity is something close to O(n log n), which I think is acceptable (and not particularly unique compared to other parts of the compiler, unless I'm missing something).

Does the explanation and conclusion look right to you?

tmandry commented on compiler/rustc_mir/src/transform/coverage/mod.rs:
```diff
@@ -129,7 +149,53 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
+                    }
+                }
+
+                ////////////////////////////////////////////////////
+                // Remove the counter or edge counter from of each `CoverageSpan`s associated
+                // `BasicCoverageBlock`, and inject a `Coverage` statement into the MIR.
+                self.inject_coverage_span_counters(
+                    coverage_spans,
+                    &mut graphviz_data,
+                    &mut debug_used_expressions,
+                );
+
+                ////////////////////////////////////////////////////
+                // For any remaining `BasicCoverageBlock` counters (that were not associated with
+                // any `CoverageSpan`), inject `Coverage` statements (_without_ code region `Span`s)
+                // to ensure `BasicCoverageBlock` counters that other `Expression`s may depend on
+                // are in fact counted, even though they don't directly contribute to counting
+                // their own independent code region's coverage.
+                self.inject_indirect_counters(&mut graphviz_data, &mut debug_used_expressions);
```
Can you say more about why things are injected in this particular order?

richkadel replied:
More comments added to explain this. Does this work?

richkadel commented:
@tmandry @wesleywiser - I think I've addressed all of the remaining feedback, but I'm re-running tests locally, first, before I upload the latest changes.

I may respond to some of the review feedback before the changes are uploaded, but they changes should be up in less than an hour, I expect.

Tyler - I added some FIXME comments to add unit tests, as you suggested. Are you OK if I defer the unit tests to a follow-on (near term) PR?

richkadel commented on compiler/rustc_mir/src/transform/coverage/counters.rs:
```diff
@@ -79,3 +110,490 @@ impl CoverageCounters {
+        Self { coverage_counters, basic_coverage_blocks }
+    }
+
+    /// If two `CoverageGraph` branch from another `BasicCoverageBlock`, one of the branches
```
Oops! Yes! This happened in a global search/replace a while back. Thanks!

richkadel commented on compiler/rustc_mir/src/transform/coverage/counters.rs:
```diff
@@ -79,3 +110,490 @@ impl CoverageCounters {
+                    bcb,
+                    self.format_counter(counter_kind),
+                );
+                counter_kind.as_operand_id()
```
Done. This change in `recursive_get_or_make_counter_operand()`, and the additional comments (per your other suggestion) make this a bit easier to read and comprehend, I think.

I did the same in `recursive_get_or_make_edge_counter_operand()` as well.

richkadel commented on compiler/rustc_mir/src/transform/coverage/counters.rs:
```diff
@@ -79,3 +110,490 @@ impl CoverageCounters {
+            bcbs_with_coverage.insert(covspan.bcb);
+        }
+
+        // FIXME(richkadel): Add more comments to explain the logic here and in the rest of this
+        // function, and refactor this function to break it up into smaller functions that are
+        // easier to understand.
```
Comments added to help explain the logic, and a FIXME comment to add unit tests, which I'd like to defer to a follow-on (but near term) PR.

richkadel commented:
@tmandry @wesleywiser - the changes are uploaded

tmandry reviewed (approved):
> Tyler - I added some FIXME comments to add unit tests, as you suggested. Are you OK if I defer the unit tests to a follow-on (near term) PR?

Yeah that's fine.

LGTM - @wesleywiser do you want to review more or should we r+?

wesleywiser commented:
@tmandry I looked though briefly but I don't think I'll have time to review this in detail for at least a few days. If you think this is good to go, don't let me hold you up!

bors commented:
:umbrella: The latest upstream changes (presumably #75534) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 36856891e230eaab0011fdb772a2f2b4e83d8a20 has been approved by `tmandry`

<!-- @bors r=tmandry 36856891e230eaab0011fdb772a2f2b4e83d8a20 -->
<!-- homu: {"type":"Approved","sha":"36856891e230eaab0011fdb772a2f2b4e83d8a20","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 36856891e230eaab0011fdb772a2f2b4e83d8a20 with merge a8cca933ed5de60b209f1d7cdbc9b78d24904ae1...
<!-- homu: {"type":"BuildStarted","head_sha":"36856891e230eaab0011fdb772a2f2b4e83d8a20","merge_sha":"a8cca933ed5de60b209f1d7cdbc9b78d24904ae1"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1344451631)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1344451631","builder_name":"checks-actions"} -->

richkadel commented:
@bors retry

bors commented:
@richkadel: :key: Insufficient privileges: not in try users

richkadel commented:
@bors retry r=tmandry

bors commented:
@richkadel: :key: Insufficient privileges: not in try users

bors commented:
@richkadel: :key: Insufficient privileges: Not in reviewers

richkadel commented:
Oops, forgot you didn't delegate to me on this PR

tmandry commented:
@bors retry

tmandry commented:
Actually there's a new commit, not sure what `retry` does in that case
@bors r+

bors commented:
:pushpin: Commit 3d7baee7378526cbb15c160a73624224d605d713 has been approved by `tmandry`

<!-- @bors r=tmandry 3d7baee7378526cbb15c160a73624224d605d713 -->
<!-- homu: {"type":"Approved","sha":"3d7baee7378526cbb15c160a73624224d605d713","approver":"tmandry"} -->

richkadel commented:
```
diff -u --strip-trailing-cr expected_show_coverage_counters.generics.txt "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/coverage-reports-deadcode/coverage-reports-deadcode"/actual_show_coverage_counters.generics.txt || \
	( grep -q '^\/\/ ignore-llvm-cov-show-diffs' ../coverage/generics.rs && \
		>&2 echo 'diff failed, but suppressed with `// ignore-llvm-cov-show-diffs` in ../coverage/generics.rs' \
	) || \
	( >&2 echo 'diff failed, and not suppressed without `// ignore-llvm-cov-show-diffs` in ../coverage/generics.rs'; \
		false \
	)
--- expected_show_coverage_counters.generics.txt	2020-11-03 03:43:00.861632884 +0000
+++ /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/coverage-reports-deadcode/coverage-reports-deadcode/actual_show_coverage_counters.generics.txt	2020-11-03 04:24:50.016590890 +0000
@@ -36,12 +36,12 @@
   10:49 -> 12:6 (count=1)
 Segment at 10:49 (count = 1), RegionEntry
 Segment at 12:6 (count = 0), Skipped
-Emitting segments for function: _RNvXs_Cs4fqI2P2rA04_8genericsINtB4_8FireworklENtNtNtCs7f2nZg1zwMz_4core3ops4drop4Drop4dropB4_
+Emitting segments for function: _RNvXs_Cs4fqI2P2rA04_8genericsINtB4_8FireworklENtNtNtCshc5W3cKmLsE_4core3ops4drop4Drop4dropB4_
 Combined regions:
   17:24 -> 19:6 (count=1)
 Segment at 17:24 (count = 1), RegionEntry
 Segment at 19:6 (count = 0), Skipped
-Emitting segments for function: _RNvXs_Cs4fqI2P2rA04_8genericsINtB4_8FireworkdENtNtNtCs7f2nZg1zwMz_4core3ops4drop4Drop4dropB4_
+Emitting segments for function: _RNvXs_Cs4fqI2P2rA04_8genericsINtB4_8FireworkdENtNtNtCshc5W3cKmLsE_4core3ops4drop4Drop4dropB4_
 Combined regions:
   17:24 -> 19:6 (count=1)
 Segment at 17:24 (count = 1), RegionEntry
```

richkadel commented:
Sadly, the demangler is apparently not applied to 
`llvm-cov`s debug output (the "counters" files).

I uploaded a change to ignore diff failures from those files because they are failing. Apparently, generic function instantiations will appear in those files,
with mangled names. 

I also added a FIXME comment, recommending we apply a demangler to the mangled function symbol names in the debug output files before comparing them. That's definitely doable but not a quick fix. Ignoring diffs for these files, in tests, is not a huge deal, since we're already comparing the final reports anyway.

richkadel commented:
@tmandry @wesleywiser - Can we submit this with rollup=never, so it doesn't break another rollup, in case any other platform-dependent issues pop up?

Thanks!

wesleywiser commented:
@bors rollup=never

richkadel commented:
> @bors rollup=never

Thanks Wesley!

Do we also need a bors retry?

It's still not in the queue.

bors commented:
@richkadel: :key: Insufficient privileges: not in try users

wesleywiser commented:
@bors r+

bors commented:
:pushpin: Commit 05992542ac022ceb49149a55cb1a1c6f4fceaf90 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 05992542ac022ceb49149a55cb1a1c6f4fceaf90 -->
<!-- homu: {"type":"Approved","sha":"05992542ac022ceb49149a55cb1a1c6f4fceaf90","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit 05992542ac022ceb49149a55cb1a1c6f4fceaf90 with merge a1edbbf95a0a93e9153d75200290cfb0db8ffc8c...
<!-- homu: {"type":"BuildStarted","head_sha":"05992542ac022ceb49149a55cb1a1c6f4fceaf90","merge_sha":"a1edbbf95a0a93e9153d75200290cfb0db8ffc8c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1350018185)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1350018185","builder_name":"checks-actions"} -->

richkadel commented:
```
thread 'rustc' panicked at 'Unexpected error creating MIR spanview HTML file: Custom { kind: NotFound, error: "IO error creating MIR dump file: \"D:/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview-base/coverage-spanview-base/mir_dump.partial_eq_counter_without_region\\\\partial_eq_counter_without_region.{impl#4}-assert_receiver_is_total_eq.-------.InstrumentCoverage.0.html\"; The system cannot find the path specified. (os error 3)" }', compiler\rustc_mir\src\transform\coverage\debug.rs:641:10
```

tmandry commented:
@bors delegate+

So you don't have to keep chasing us down :)

bors commented:
:v: @richkadel can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"tmandry","delegate":"richkadel"} -->

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit fc56c6f725b9593bfe4e4216b1df46493cc435f0 has been approved by `tmandry`

<!-- @bors r=tmandry fc56c6f725b9593bfe4e4216b1df46493cc435f0 -->
<!-- homu: {"type":"Approved","sha":"fc56c6f725b9593bfe4e4216b1df46493cc435f0","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit fc56c6f725b9593bfe4e4216b1df46493cc435f0 with merge f6f8e0c44854a0e89d940c2da9e491ea4b915121...
<!-- homu: {"type":"BuildStarted","head_sha":"fc56c6f725b9593bfe4e4216b1df46493cc435f0","merge_sha":"f6f8e0c44854a0e89d940c2da9e491ea4b915121"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1351363391)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1351363391","builder_name":"checks-actions"} -->

richkadel commented:
Spurious network error:

```
Command failed. Attempt 5/5:
    Updating crates.io index
warning: spurious network error (2 tries remaining): error inflating zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error inflating zlib stream; class=Zlib (5)
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  error inflating zlib stream; class=Zlib (5)
failed to run: D:\a\rust\rust\build\i686-pc-windows-gnu\stage0\bin\cargo.exe build --manifest-path D:\a\rust\rust\src/bootstrap/Cargo.toml --locked
Build completed unsuccessfully in 0:00:22
make: *** [Makefile:60: prepare] Error 1
The command has failed after 5 attempts.
```

richkadel commented:
@bors retry

bors commented:
:hourglass: Testing commit fc56c6f725b9593bfe4e4216b1df46493cc435f0 with merge 6d1f6bffb9b236f40fe5c59a604352e49d47f44a...
<!-- homu: {"type":"BuildStarted","head_sha":"fc56c6f725b9593bfe4e4216b1df46493cc435f0","merge_sha":"6d1f6bffb9b236f40fe5c59a604352e49d47f44a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1351567734)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1351567734","builder_name":"checks-actions"} -->

richkadel commented:
`llvm-cov show` mysteriously failed on Windows, but because I was capturing stderr to get the debug counters output, I was suppressing the error message. 

I don't know why it failed. Perhaps its spurious, but since it failed on the first sample in the test suite, I doubt it.

Testing again so, if it does still fail, I'll at least get an error message to help deduce the issue.

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit 1ce58271729d6f86c1d62f2d311fa0ab67d6f2e3 has been approved by `tmandry`

<!-- @bors r=tmandry 1ce58271729d6f86c1d62f2d311fa0ab67d6f2e3 -->
<!-- homu: {"type":"Approved","sha":"1ce58271729d6f86c1d62f2d311fa0ab67d6f2e3","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 1ce58271729d6f86c1d62f2d311fa0ab67d6f2e3 with merge fa80967e81fac3610c83287f26f0fbee8baf755d...
<!-- homu: {"type":"BuildStarted","head_sha":"1ce58271729d6f86c1d62f2d311fa0ab67d6f2e3","merge_sha":"fa80967e81fac3610c83287f26f0fbee8baf755d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1352002435)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1352002435","builder_name":"checks-actions"} -->

richkadel commented:
```
------------------------------------------
stderr:
------------------------------------------
llvm-cov.exe show: Unknown command line argument '--debug'.  Try: 'D:\a\rust\rust\build\x86_64-pc-windows-msvc\llvm\build\bin\llvm-cov.exe show --help'
llvm-cov.exe show: Did you mean '--dump'?
```

Wait... what?

OK, well I don't need the counters file on Windows.

I'm checking out the command on Windows locally, to see if maybe it has a different flag for this. Otherwise, I'll change the Makefile to remove the `--debug` flag and ignore diffs on Windows.

richkadel commented:
OK, it looks like this is CI specific. From the llvm-cov source, it **appears** that the `--debug` flag is only enabled when the binary is build with debug assertions enabled.

I took a look at `src/ci/github-actions/ci.yml`, and all of the Windows (both MSVC and gnu) build configurations include `env` settings:

* `NO_LLVM_ASSERTIONS: 1`
* `NO_DEBUG_ASSERTIONS: 1`

Along with an accompanying FIXME(#59637).

Interestingly, the assertions are also disabled in `apple` builds, but without a `FIXME` comment, unfortunately.

I'll reference the `FIXME` for Windows in my Makefile edits, but to be safe, I think I will only enable the `--debug` flag and check the counters under Linux builds.

richkadel commented:
Better yet, IIUC, the `env` settings are being set in the main CI shell environment, and are used in `src/ci/run.sh` to set command line flags for the build, but not cleared or anything; so it seems to me I can reference these env variable values in the test (Makefile) rather that making it platform dependent.

That way, if they are re-enabled, or if the get disabled in a Linux distro, the Makefile should still work.

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit 0d00fd9c125b791ae6f02b8142e6e172f56849ca has been approved by `tmandry`

<!-- @bors r=tmandry 0d00fd9c125b791ae6f02b8142e6e172f56849ca -->
<!-- homu: {"type":"Approved","sha":"0d00fd9c125b791ae6f02b8142e6e172f56849ca","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 0d00fd9c125b791ae6f02b8142e6e172f56849ca with merge 414423ee00f58990f1072a7ff92f0e40585db4e0...
<!-- homu: {"type":"BuildStarted","head_sha":"0d00fd9c125b791ae6f02b8142e6e172f56849ca","merge_sha":"414423ee00f58990f1072a7ff92f0e40585db4e0"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1356495074)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1356495074","builder_name":"checks-actions"} -->

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90 has been approved by `tmandry`

<!-- @bors r=tmandry c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90 -->
<!-- homu: {"type":"Approved","sha":"c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90 with merge b65be804bdf38b4d9093c1333b6f253c4171eabf...
<!-- homu: {"type":"BuildStarted","head_sha":"c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90","merge_sha":"b65be804bdf38b4d9093c1333b6f253c4171eabf"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1357671801)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1357671801","builder_name":"checks-actions"} -->

richkadel commented:
Failure was not related to this PR? On apple-x86_64:

```

test [debuginfo-lldb] debuginfo/pretty-std-collections-hash.rs ... ok
test [debuginfo-lldb] debuginfo/pretty-std-collections.rs ... FAILED
test [debuginfo-lldb] debuginfo/pretty-std.rs ... ignored
test [debuginfo-lldb] debuginfo/pretty-uninitialized-vec.rs ... ok

...


test [debuginfo-lldb] debuginfo/var-captured-in-stack-closure.rs ... ok
test [debuginfo-lldb] debuginfo/vec-slices.rs ... ok
thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:354:22
test [debuginfo-lldb] debuginfo/vec.rs ... ok
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

failures:

---- [debuginfo-lldb] debuginfo/pretty-std-collections.rs stdout ----
NOTE: compiletest thinks it is using LLDB version 1200
NOTE: compiletest thinks it is using LLDB without native rust support

error: Error while running LLDB
status: exit code: 1
command: "/usr/bin/python3" "/Users/runner/work/rust/rust/src/etc/lldb_batchmode.py" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/a" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/pretty-std-collections.debugger.script"
stdout:
------------------------------------------
LLDB batch-mode script
----------------------
Debugger commands script is '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/pretty-std-collections.debugger.script'.
Target executable is '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/a'.
Current working directory is '/Users/runner/work/rust/rust'
Creating a target for '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/a'
settings set auto-confirm true

version
lldb-1200.0.32.1 Apple Swift version 5.3 (swiftlang-1200.0.29.2 clang-1200.0.30.1) 
command script import /Users/runner/work/rust/rust/./src/etc/lldb_lookup.py
type synthetic add -l lldb_lookup.synthetic_lookup -x '.*' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)String$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^&str$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^&\[.+\]$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::ffi::([a-z_]+::)+)OsString$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Vec<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)VecDeque<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)BTreeSet<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)BTreeMap<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::collections::([a-z_]+::)+)HashMap<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::collections::([a-z_]+::)+)HashSet<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Rc<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Arc<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)Cell<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)Ref<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)RefMut<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)RefCell<.+>$' --category Rust
type category enable Rust

breakpoint set --file 'pretty-std-collections.rs' --line 158
DEBUG: breakpoint added, id = 1
Breakpoint 1: where = a`pretty_std_collections::main::h1ef1d7ce5967b6a7 + 1513 at pretty-std-collections.rs:158:5, address = 0x0000000100029399 
DEBUG: registering breakpoint callback, id = 1
Error while trying to register breakpoint callback, id = 1, message = error: could not get num args: can't find callable: breakpoint_callback

run
Process 65303 stopped * thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1 frame #0: 0x0000000100029399 a`pretty_std_collections::main::h1ef1d7ce5967b6a7 at pretty-std-collections.rs:158:5 155 hash_set.insert(i); 156 } 157 -> 158 zzz(); // #break ^ 159 } 160 161 fn zzz() { Target 0: (a) stopped. Process 65303 launched: '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/a' (x86_64) 
print vec_deque
(alloc::collections::vec_deque::VecDeque<int>) $0 = size=3 { [0] = 5 [1] = 3 [2] = 7 } 
print vec_deque2
(alloc::collections::vec_deque::VecDeque<int>) $1 = size=7 { [0] = 2 [1] = 3 [2] = 4 [3] = 5 [4] = 6 [5] = 7 [6] = 8 } 
print hash_map

------------------------------------------
stderr:
------------------------------------------
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
clang: CommandLine Error: Option 'h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

------------------------------------------



failures:
    [debuginfo-lldb] debuginfo/pretty-std-collections.rs

test result: FAILED. 89 passed; 1 failed; 26 ignored; 0 measured; 0 filtered out
```

wesleywiser commented:
That's #78665. We can just retry.

@bors retry

richkadel commented:
The good news is, CI passed on all platforms except apple (in the last 2 attempts), and it was clear why it failed on apple, and should be fixed. I thinks its very likely to pass if not hit with another spurious failure.

bors commented:
:hourglass: Testing commit c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90 with merge c3aa35358c2e924ea6e0dd595e4c92cea0cef909...
<!-- homu: {"type":"BuildStarted","head_sha":"c4334b63cb3d27ab67778b1c2b3cfc5ca7bbda90","merge_sha":"c3aa35358c2e924ea6e0dd595e4c92cea0cef909"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1361524153)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1361524153","builder_name":"checks-actions"} -->

richkadel commented:
The sed command breaks on MacOS, and the simple fix will likely produce inconsistent results. I'm disabling this on MacOS for now.

richkadel commented:
@bors retry r=tmandry

bors commented:
:pushpin: Commit ece2792f1dab759706a2f04c75aeeb8ad2deb5ac has been approved by `tmandry`

<!-- @bors r=tmandry ece2792f1dab759706a2f04c75aeeb8ad2deb5ac -->
<!-- homu: {"type":"Approved","sha":"ece2792f1dab759706a2f04c75aeeb8ad2deb5ac","approver":"tmandry"} -->

richkadel commented:
@bors retry r=tmandry

bors commented:
:pushpin: Commit bf0f0425ad73b27b84ae549cd0a2ece1e9769510 has been approved by `tmandry`

<!-- @bors r=tmandry bf0f0425ad73b27b84ae549cd0a2ece1e9769510 -->
<!-- homu: {"type":"Approved","sha":"bf0f0425ad73b27b84ae549cd0a2ece1e9769510","approver":"tmandry"} -->

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit 68014e6b1ceb34da29b51a2804779138c6275fa4 has been approved by `tmandry`

<!-- @bors r=tmandry 68014e6b1ceb34da29b51a2804779138c6275fa4 -->
<!-- homu: {"type":"Approved","sha":"68014e6b1ceb34da29b51a2804779138c6275fa4","approver":"tmandry"} -->

richkadel commented:
I found a typo in my MacOS fix and resolved it, and then found that --debug in llvm-cov show doesn't work even when developing outside of CI. (The CI should pass now, but without disabling it for all MacOS builds, it would have failed for developers.) Took care of that as well.

bors commented:
:hourglass: Testing commit 68014e6b1ceb34da29b51a2804779138c6275fa4 with merge 8532e742fc6ec210fab69b8192190bc40c685912...
<!-- homu: {"type":"BuildStarted","head_sha":"68014e6b1ceb34da29b51a2804779138c6275fa4","merge_sha":"8532e742fc6ec210fab69b8192190bc40c685912"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1362842586)
Approved by: tmandry
Pushing 8532e742fc6ec210fab69b8192190bc40c685912 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"tmandry","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1362842586"},"merge_sha":"8532e742fc6ec210fab69b8192190bc40c685912"} -->
